### PR TITLE
Reworking doc hover

### PR DESCRIPTION
**Pull request by sellophane**

### CODE REVIEW

File: kotlin-eclipse-core/src/org/jetbrains/kotlin/core/preferences/KotlinProperties.kt [hunk 1]
```diff
@@ -13,7 +13,10 @@ class KotlinProperties(scope: IScopeContext = InstanceScope.INSTANCE) : Preferen
     var globalsOverridden by BooleanPreference()
 
     // Note: default value is defined in preferences.ini
-    var jvmTarget by EnumPreference<JvmTarget>(JvmTarget.DEFAULT)
+    var jvmTarget by object : Preference<JvmTarget> {
+    	override fun reader(text: String?) = text?.let { enumValueOf<JvmTarget>(it) } ?: JvmTarget.DEFAULT
+        override fun writer(value: JvmTarget) = value.name
+	}
```
Comment (zarechenskiy) [inline kotlin-eclipse-core/src/org/jetbrains/kotlin/core/preferences/KotlinProperties.kt]: Minor: formatting `}`

File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt [hunk 2]
```diff
@@ -16,52 +16,198 @@
...
-class KotlinJavadocTextHover : KotlinEditorTextHover<BrowserInformationControlInput> {
+class KotlinJavadocTextHover : KotlinEditorTextHover<BrowserInformationControlInput>, ITextHoverExtension {
+
+    class KotlinInformationControlCreator : IInformationControlCreator {
+
+        override fun createInformationControl(parent: Shell?): IInformationControl =
+            if (BrowserInformationControl.isAvailable(parent)) {
+                val tbm = ToolBarManager(8388608)
+                val font = "org.eclipse.jdt.ui.javadocfont"
+                BrowserInformationControl(parent, font, tbm)
```
Comment (zarechenskiy) [inline kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt]: `tbm` and `font` are constants, so let's extract them to a `companion` object, for example. Plus, please explain why `8388608` is used

File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt [hunk 3]
```diff
@@ -16,52 +16,198 @@
...
-        val (element, editor) = hoverData
-        
-        val javaElements = obtainJavaElements(element)
-        if (javaElements.isEmpty()) return null
-        
-        val editorInputElement = EditorUtility.getEditorInputJavaElement(editor.javaEditor, false)
-        
-        return JavadocHover.getHoverInfo(javaElements.toTypedArray(), editorInputElement, DUMMY_REGION, null)
+        val (element, _) = hoverData
+        return KotlinAnnotationBrowserInformationControlInput(element)
```
Comment (zarechenskiy) [inline kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt]: No reason to use additional `val`: `KotlinAnnotationBrowserInformationControlInput(hoverData.hoverElement)`

File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt [hunk 4]
```diff
@@ -16,52 +16,198 @@
...
+
+            fun img(src: URL) =
+                "<img style='border:none; position: absolute; width: ${imageWidth}px; height: ${imageHeight}px; left: -${labelLeft + 1}px;' src='$src'/>"
+        }
+
+        private val expression = getReferenceExpression(element)
+
+        private val descriptor = expression?.let {
+            KotlinAnalysisFileCache.getAnalysisResult(element.containingKtFile).analysisResult.bindingContext[REFERENCE_TARGET, expression]
+        } ?: (element as? KtDeclaration)?.resolveToDescriptor()
```
Comment (zarechenskiy) [inline kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt]: Seems quite complicated to me, why not just 
```
(expression ?: element as? KtDeclaration)?.resolveToDescriptor()
```
?

**Comment by zarechenskiy** [comment]
@Kordyjan Could you please also review these changes?

File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt (hunk 4 above)
Comment (sellophane) [reply kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt]: Actually, resolveToDescriptor browses DECLARATION_TO_DESCRIPTOR Slice and I need to check REFERENCE_TARGET one instead.

File: kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/navigation/navigationUtils.kt [hunk 5]
```diff
@@ -295,6 +296,22 @@ private fun tryToFindExternalClassInStdlib(
     return openKotlinEditorForExternalFile(content, name, pckg, ktFile)
 }
 
+fun getKtFileFromElement(element: KtElement, descriptor: DeclarationDescriptor): KtFile? {
```
Comment (Kordyjan) [inline kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/navigation/navigationUtils.kt]: isn't it something more like `getSourceFileForElement`?

**Pull request merged**